### PR TITLE
Fix one/many vs. raw benchmark output for a number of formats

### PR DESCRIPTION
**Pull request by magnumripper**

### CODE REVIEW

**Comment by solardiz** [comment]
I think your changes for andotp, blockchain, and gost are wrong - please re-read my comments on those formats in #3795.

**Comment by magnumripper** [comment]
Oh, I missed that, I was only looking at the comments to d99065fd plus new benchmark output analysis by fixed perl script. The script now complains about 0x400 when costs doesn't differ. I'll add comments to these formats while at it.

**Comment by magnumripper** [comment]
Fixed now - although a comment fix to gost is missing: I added a comment to andotp (actually mentioned in the thread) and blockchain (reason mentioned in a FIXME comment in 4cbd6405) but can't find any information on why gost need this so I just added `// FIXME document why!` to that one.

**Comment by solardiz** [comment]
"gost" is actually unsalted, and can only have at most 2 different "salts" for two variations of the algorithm. So our "Many salts" benchmark simulating up to 256 different salts is testing an impossible condition.

**Comment by solardiz** [comment]
I'll double-check a few more of these changes - please don't merge yet.

File: src/o3logon_fmt_plug.c [hunk 1]
```diff
@@ -41,7 +41,7 @@ john_register_one(&fmt_o3logon);
 #define ALGORITHM_NAME                  "SHA1 DES 32/" ARCH_BITS_STR
 
 #define BENCHMARK_COMMENT               ""
-#define BENCHMARK_LENGTH                0x107
+#define BENCHMARK_LENGTH                7
```
Comment (solardiz) [inline src/o3logon_fmt_plug.c]: OK, but I am puzzled why "Many salts" gets limited to as few as 3 salts when the speed is rather high (about as high as for o10glogon above, where we got ~150 salts in a similar test). I guess very different MKPC*OMP_SCALE, but why would optimal be very different for formats that are of similar speed? Confirmed: o10glogon has `16*8`, whereas o3logon has `64*128`. One or both of these are probably mistuned.

```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=o3logon  
Will run 256 OpenMP threads
Benchmarking: o3logon, Oracle O3LOGON protocol [SHA1 DES 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 3/256
Many salts:     4279K c/s real, 17453 c/s virtual
Only one salt:  2150K c/s real, 17568 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=o3logon
Will run 256 OpenMP threads
Benchmarking: o3logon, Oracle O3LOGON protocol [SHA1 DES 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 3/256
Many salts:     4309K c/s real, 17463 c/s virtual
Only one salt:  2129K c/s real, 17557 c/s virtual
```

File: src/o10glogon_fmt_plug.c [hunk 2]
```diff
@@ -45,7 +45,7 @@ john_register_one(&fmt_o10glogon);
 #define ALGORITHM_NAME                  "DES-AES128-MD5 32/" ARCH_BITS_STR
 
 #define BENCHMARK_COMMENT               ""
-#define BENCHMARK_LENGTH                0x107
+#define BENCHMARK_LENGTH                7
```
Comment (solardiz) [inline src/o10glogon_fmt_plug.c]: OK:
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=o10glogon
Will run 256 OpenMP threads
Benchmarking: o10glogon, Oracle 10g-logon protocol [DES-AES128-MD5 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 147/256
Many salts:     4816K c/s real, 21989 c/s virtual
Only one salt:  2141K c/s real, 21015 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=o10glogon
Will run 256 OpenMP threads
Benchmarking: o10glogon, Oracle 10g-logon protocol [DES-AES128-MD5 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 149/256
Many salts:     4834K c/s real, 21977 c/s virtual
Only one salt:  2097K c/s real, 20950 c/s virtual
```

File: src/ospf_fmt_plug.c [hunk 3]
```diff
@@ -39,7 +39,7 @@ john_register_one(&fmt_ospf);
 #define TAG_LENGTH              (sizeof(FORMAT_TAG) - 1)
 #define ALGORITHM_NAME          "HMAC-SHA-X 32/" ARCH_BITS_STR
 #define BENCHMARK_COMMENT       ""
-#define BENCHMARK_LENGTH        0x107
+#define BENCHMARK_LENGTH        7
```
Comment (solardiz) [inline src/ospf_fmt_plug.c]: This looked OK at first, but no. The first two test vectors are different speed. Please use `0x507` here and add a comment. Perhaps also add a FIXME comment requesting reporting of tunable costs.
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=ospf   
Will run 256 OpenMP threads
Benchmarking: ospf, OSPF / IS-IS [HMAC-SHA-X 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 7/256
Many salts:     12336K c/s real, 53119 c/s virtual
Only one salt:  8962K c/s real, 81945 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=ospf
Will run 256 OpenMP threads
Benchmarking: ospf, OSPF / IS-IS [HMAC-SHA-X 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 7/256
Many salts:     12440K c/s real, 53136 c/s virtual
Only one salt:  9039K c/s real, 82003 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ./john -mask -max-run-time=10 pw-ospf1
Using default input encoding: UTF-8
Loaded 1 password hash (ospf, OSPF / IS-IS [HMAC-SHA-X 32/64])
Will run 256 OpenMP threads
Using default mask: ?1?2?2?2?2?2?2?3?3?3?3?d?d?d?d
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:10  0g/s 6738Kp/s 6738Kc/s 6738KC/s ncc4zaaaaaa1111..r8hzxaaaaaa1111
Session stopped (max run-time reached)
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ./john -mask -max-run-time=10 pw-ospf2
Using default input encoding: UTF-8
Loaded 1 password hash (ospf, OSPF / IS-IS [HMAC-SHA-X 32/64])
Will run 256 OpenMP threads
Using default mask: ?1?2?2?2?2?2?2?3?3?3?3?d?d?d?d
Press 'q' or Ctrl-C to abort, almost any other key for status
0g 0:00:00:10  0g/s 5075Kp/s 5075Kc/s 5075KC/s Hfgnpaaaaaa1111..Ulf9paaaaaa1111
Session stopped (max run-time reached)
[solar@localhost run]$ cat pw-ospf1
$ospf$1$02010030ac10001400000000000000020000011454ee4518ffffff00000a120100000028c0a86f14c0a86f0aac10000a$e59ba2c56a2c0429ebe72a194e4b54c250cac1a3
[solar@localhost run]$ cat pw-ospf2
$ospf$2$0201002cac10000a00000000000000020000012054f4c8adffffff00000a120100000028c0a86f0a00000000$508a1abffb5b4554e1aa46eb053bca7105c3e8f6fece4c945f0a0020edb054ec
```

File: src/pbkdf2_hmac_sha256_fmt_plug.c [hunk 4]
```diff
@@ -43,7 +43,6 @@ john_register_one(&fmt_pbkdf2_hmac_sha256);
 
 #define FORMAT_LABEL            "PBKDF2-HMAC-SHA256"
 #define FORMAT_NAME		""
-#define BENCHMARK_LENGTH        0x107
```
Comment (solardiz) [inline src/pbkdf2_hmac_sha256_fmt_plug.c]: I guess this was a redefine of the same value coming from elsewhere?

```c
struct fmt_tests pbkdf2_hmac_sha256_common_tests[] = {
        /* Low iteration test vectors for comparison */
        {"$pbkdf2-sha256$1000$b1dWS2dab3dKQWhPSUg3cg$UY9j5wlyxtsJqhDKTqua8Q3fMp0ojc2pOnErzr8ntLE", "magnum"},
        {"$pbkdf2-sha256$1000$amkzNk9tOXJVZ043QTZ1dA$YIpbl0hjiV9UFMszHNDlJIa0bcObTIkPzT6XvhJPcMM", "bonum"},
```

Looks OK so far.

```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=pbkdf2-hmac-sha256
Will run 256 OpenMP threads
Benchmarking: PBKDF2-HMAC-SHA256 [PBKDF2-SHA256 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 1000
Warning: "Many salts" test limited: 14/256
Many salts:     449756 c/s real, 1865 c/s virtual
Only one salt:  436906 c/s real, 1867 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=pbkdf2-hmac-sha256
Will run 256 OpenMP threads
Benchmarking: PBKDF2-HMAC-SHA256 [PBKDF2-SHA256 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 1000
Warning: "Many salts" test limited: 14/256
Many salts:     441107 c/s real, 1865 c/s virtual
Only one salt:  436906 c/s real, 1870 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=pbkdf2-hmac-sha256
Will run 256 OpenMP threads
Benchmarking: PBKDF2-HMAC-SHA256 [PBKDF2-SHA256 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 1000
Warning: "Many salts" test limited: 14/256
Many salts:     445390 c/s real, 1865 c/s virtual
Only one salt:  428740 c/s real, 1867 c/s virtual
```

Not sure why the difference, but perhaps it's just the sequential `set_key` overhead. This difference is much smaller than md5crypt's on this same machine, so I'm OK with 0x107.

**Comment by magnumripper** [comment]
I pushed a separate commit fixing the GOST comment (so we'll squash & merge this PR, please do so if/when you've reviewed)

File: src/pdf_fmt_plug.c [hunk 5]
```diff
@@ -40,7 +40,7 @@ john_register_one(&fmt_pdf);
 #define FORMAT_TAG_OLD_LEN  (sizeof(FORMAT_TAG_OLD)-1)
 #define ALGORITHM_NAME      "MD5 SHA2 RC4/AES 32/" ARCH_BITS_STR
 #define BENCHMARK_COMMENT   ""
-#define BENCHMARK_LENGTH    7
+#define BENCHMARK_LENGTH    0x507
```
Comment (solardiz) [inline src/pdf_fmt_plug.c]: OK. Different costs, reported as such. Please add a comment about that.

File: src/postgres_fmt_plug.c [hunk 6]
```diff
@@ -44,7 +44,7 @@ john_register_one(&fmt_postgres);
 #define FORMAT_TAG2_LEN         (sizeof(FORMAT_TAG2)-1)
 #define ALGORITHM_NAME          "MD5 32/" ARCH_BITS_STR
 #define BENCHMARK_COMMENT       ""
-#define BENCHMARK_LENGTH        0x107
+#define BENCHMARK_LENGTH        7
```
Comment (solardiz) [inline src/postgres_fmt_plug.c]: OK:
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=postgres
Will run 256 OpenMP threads
Benchmarking: postgres, PostgreSQL C/R [MD5 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 203/256
Many salts:     26607K c/s real, 160122 c/s virtual
Only one salt:  11534K c/s real, 157422 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=postgres
Will run 256 OpenMP threads
Benchmarking: postgres, PostgreSQL C/R [MD5 32/64]... (256xOMP) DONE
Warning: "Many salts" test limited: 205/256
Many salts:     26603K c/s real, 156931 c/s virtual
Only one salt:  11809K c/s real, 158148 c/s virtual
```

File: src/sapH_fmt_plug.c [hunk 7]
```diff
@@ -84,7 +84,7 @@ john_register_one(&fmt_sapH);
 
 
 #define BENCHMARK_COMMENT		" (SHA1x1024)"
-#define BENCHMARK_LENGTH		0x507
+#define BENCHMARK_LENGTH		7
```
Comment (solardiz) [inline src/sapH_fmt_plug.c]: OK:
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=saph    
Will run 256 OpenMP threads
Benchmarking: saph, SAP CODVN H (PWDSALTEDHASH) (SHA1x1024) [SHA-1/SHA-2 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (hash type [1:SHA1 2:SHA256 3:SHA384 4:SHA512]) of 1, cost 2 (iteration count) of 1024
Warning: "Many salts" test limited: 22/256
Many salts:     1441K c/s real, 6067 c/s virtual
Only one salt:  1323K c/s real, 6061 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=saph
Will run 256 OpenMP threads
Benchmarking: saph, SAP CODVN H (PWDSALTEDHASH) (SHA1x1024) [SHA-1/SHA-2 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (hash type [1:SHA1 2:SHA256 3:SHA384 4:SHA512]) of 1, cost 2 (iteration count) of 1024
Warning: "Many salts" test limited: 22/256
Many salts:     1427K c/s real, 6071 c/s virtual
Only one salt:  1336K c/s real, 6075 c/s virtual
```

File: src/solarwinds_fmt_plug.c [hunk 8]
```diff
@@ -50,7 +50,7 @@ john_register_one(&fmt_solarwinds);
 #define ALGORITHM_NAME       "PBKDF2-SHA1 32/" ARCH_BITS_STR
 #endif
 #define BENCHMARK_COMMENT    ""
-#define BENCHMARK_LENGTH     0x107
+#define BENCHMARK_LENGTH     7
```
Comment (solardiz) [inline src/solarwinds_fmt_plug.c]: Perhaps you had a spurious result here, please revert.
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=solarwinds
Will run 256 OpenMP threads
Benchmarking: solarwinds, SolarWinds Orion [PBKDF2-SHA1 512/512 AVX512F 16x]... (256xOMP) DONE
Warning: "Many salts" test limited: 5/256
Many salts:     20277 c/s real, 81.5 c/s virtual
Only one salt:  20277 c/s real, 81.6 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=solarwinds
Will run 256 OpenMP threads
Benchmarking: solarwinds, SolarWinds Orion [PBKDF2-SHA1 512/512 AVX512F 16x]... (256xOMP) DONE
Warning: "Many salts" test limited: 5/256
Many salts:     20078 c/s real, 81.6 c/s virtual
Only one salt:  20078 c/s real, 81.6 c/s virtual
```

File: src/sspr_fmt_plug.c [hunk 9]
```diff
@@ -38,7 +38,7 @@ john_register_one(&fmt_sspr);
 #define FORMAT_LABEL            "sspr"
 #define ALGORITHM_NAME          "MD5/SHA1/SHA256/SHA512 32/" ARCH_BITS_STR
 #define BENCHMARK_COMMENT       ""
-#define BENCHMARK_LENGTH        0x507
+#define BENCHMARK_LENGTH        0x107
```
Comment (solardiz) [inline src/sspr_fmt_plug.c]: OK. I am getting same many vs. one speeds.

File: src/telegram_fmt_plug.c [hunk 10]
```diff
@@ -45,7 +45,7 @@ john_register_one(&fmt_telegram);
 #define ALGORITHM_NAME          "PBKDF2-SHA1 32/" ARCH_BITS_STR " AES"
 #endif
 #define BENCHMARK_COMMENT       ""
-#define BENCHMARK_LENGTH        0x107
+#define BENCHMARK_LENGTH        7
```
Comment (solardiz) [inline src/telegram_fmt_plug.c]: Perhaps you had a spurious result here, please revert.
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=telegram
Will run 256 OpenMP threads
Benchmarking: telegram [PBKDF2-SHA1 512/512 AVX512F 16x AES]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 4000
Warning: "Many salts" test limited: 3/256
Many salts:     37809 c/s real, 151 c/s virtual
Only one salt:  37520 c/s real, 151 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=telegram
Will run 256 OpenMP threads
Benchmarking: telegram [PBKDF2-SHA1 512/512 AVX512F 16x AES]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 4000
Warning: "Many salts" test limited: 3/256
Many salts:     37236 c/s real, 151 c/s virtual
Only one salt:  37809 c/s real, 151 c/s virtual
```

File: src/vmx_fmt_plug.c [hunk 11]
```diff
@@ -41,7 +41,7 @@ john_register_one(&fmt_vmx);
 #define ALGORITHM_NAME          "PBKDF2-SHA1 AES 32/" ARCH_BITS_STR
 #endif
 #define BENCHMARK_COMMENT       ""
-#define BENCHMARK_LENGTH        0x107
+#define BENCHMARK_LENGTH        7
```
Comment (solardiz) [inline src/vmx_fmt_plug.c]: Perhaps you had a spurious result here, please revert.
```
[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=vmx     
Will run 256 OpenMP threads
Benchmarking: vmx, VMware VMX [PBKDF2-SHA1 AES 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 10000
Warning: "Many salts" test limited: 13/256
Many salts:     51200 c/s real, 213 c/s virtual
Only one salt:  51697 c/s real, 213 c/s virtual

[solar@localhost run]$ GOMP_CPU_AFFINITY=0-255 GOMP_SPINCOUNT=10000 ../run/john -test -form=vmx
Will run 256 OpenMP threads
Benchmarking: vmx, VMware VMX [PBKDF2-SHA1 AES 512/512 AVX512F 16x]... (256xOMP) DONE
Speed for cost 1 (iteration count) of 10000
Warning: "Many salts" test limited: 13/256
Many salts:     51697 c/s real, 213 c/s virtual
Only one salt:  51697 c/s real, 213 c/s virtual
```

**Comment by solardiz** [comment]
While testing these formats, I noticed that "Many salts" sometimes gets limited to 1 salt - perhaps we should revise that code to limit to no fewer than 2 salts.

File: src/bestcrypt_fmt_plug.c [hunk 12]
```diff
@@ -46,7 +46,7 @@ john_register_one(&fmt_bestcrypt);
 #define BINARY_SIZE             0
 #define BINARY_ALIGN            1
 #define BENCHMARK_COMMENT       " (SHA-256 + AES XTS mode)"
-#define BENCHMARK_LENGTH        0x507
+#define BENCHMARK_LENGTH        7
```
[the remark this answers is not in the document]
Comment (magnumripper) [reply src/bestcrypt_fmt_plug.c]: Oh. So the format lacks a cost definition?

File: src/episerver_fmt_plug.c [hunk 13]
```diff
@@ -62,7 +62,7 @@ john_register_one(&fmt_episerver);
 #define FORMAT_TAG              "$episerver$*"
 #define FORMAT_TAG_LEN          (sizeof(FORMAT_TAG)-1)
 #define BENCHMARK_COMMENT       ""
-#define BENCHMARK_LENGTH        7
+#define BENCHMARK_LENGTH        0x107
```
[the remark this answers is not in the document]
Comment (magnumripper) [reply src/episerver_fmt_plug.c]: I got this from super (using `--test=-1`)
```
Benchmarking: EPiServer [SHA1/SHA256 128/128 AVX 4x]... (32xOMP) DONE
Speed for cost 1 (hash type [1:SHA1 2:SHA256]) of 1
Many salts:	33882K c/s real, 5277K c/s virtual
Only one salt:	41287K c/s real, 5376K c/s virtual
```
But I admit I didn't test it manually. Will fix.

File: src/pbkdf2_hmac_sha256_fmt_plug.c (hunk 4 above)
Comment (magnumripper) [reply src/pbkdf2_hmac_sha256_fmt_plug.c]: Yup, it's in `pbkdf2_hmac_common.h`. I got spurious reports of barely reaching 1 or 2% from a few of the formats but decided we'll just go with raw for all of them.

File: src/pdf_fmt_plug.c (hunk 5 above)
Comment (magnumripper) [reply src/pdf_fmt_plug.c]: For now I have reported so for any formats *not* defining them.

**Comment by magnumripper** [comment]
All fixed

**Pull request merged**